spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/BaseDataReader.java

```diff
@@ -60,35 +64,54 @@
  *
  * @param <T> is the Java class returned by this reader whose objects contain one or more rows.
  */
-abstract class BaseDataReader<T> implements Closeable {
+abstract class BaseDataReader<T, CST extends ContentScanTask<?>, G extends ScanTaskGroup<CST>>
```

Comment:
To be honest, I am not convinced we have to restrict this to `ContentScanTask`. Also, keep in mind that different changelog tasks can be packed into the same task group. That's why the reader should not be restricted to strictly one task type.

I'd consider using `ScanTask` and changing the hierarchy a bit. This can can become `BaseReader`. I also don't think we need the second type parameter. We can just work with `ScanTaskGroup<ST>`.

```
abstract class BaseReader<T, ST extends ScanTask> implements Closeable {

  ...

  BaseReader(Table table, ScanTaskGroup<ST> taskGroup) {
    this.table = table;
    this.tasks = taskGroup.tasks().iterator();
    this.inputFiles = inputFiles(taskGroup);
    this.currentIterator = CloseableIterator.empty();
  }

  ...
}
```

Then I'd consider adding `BaseRowReader` like this (also no data in the name).

```
abstract class BaseRowReader<ST extends ScanTask> extends BaseReader<InternalRow, ST> {

  private final Schema tableSchema;
  private final Schema expectedSchema;
  private final String nameMapping;
  private final boolean caseSensitive;

  BaseRowReader(Table table, ScanTaskGroup<ST> taskGroup, Schema expectedSchema, boolean caseSensitive) {
    super(table, taskGroup);
    this.tableSchema = table.schema();
    this.expectedSchema = expectedSchema;
    this.nameMapping = table.properties().get(TableProperties.DEFAULT_NAME_MAPPING);
    this.caseSensitive = caseSensitive;
  }

  protected Schema tableSchema() {
    return tableSchema;
  }

  protected Schema expectedSchema() {
    return expectedSchema;
  }

  protected CloseableIterable<InternalRow> newIterable(InputFile file, FileFormat format, long start, long length,
                                                       Expression residual, Schema projection,
                                                       Map<Integer, ?> idToConstant) {
    switch (format) {
      case PARQUET:
        return newParquetIterable(file, start, length, residual, projection, idToConstant);

      case AVRO:
        return newAvroIterable(file, start, length, projection, idToConstant);

      case ORC:
        return newOrcIterable(file, start, length, residual, projection, idToConstant);

      default:
        throw new UnsupportedOperationException("Cannot read unknown format: " + format);
    }
  }

  ...
}
```

Then the existing `RowDataReader` won't need to change a lot.

```
class RowDataReader extends BaseRowReader<FileScanTask> {
  ...
}
```

Finally, we will have `ChangelogRowReader` capable of reading all types of changelog tasks like this.

```
class ChangelogRowReader extends BaseRowReader<ChangelogScanTask> {
  ...
}
```

Reply:
I agree that the second type parameter isn't needed. Thanks for the suggestion.

If we want to use `ScanTask` in the `BaseReader/BaseDataReader`, we need to either 
1. move some common logic to its subclasses, e.g. `getInputFile()`, `constantsMap()` , which need the method from `ContentScanTask`.
2. add new methods in `ScanTask`. e.g.  `file()`